src/syscalls_new.rs

```diff
@@ -0,0 +1,190 @@
+    // clobber any of the other caller-saved registers. Thus we mark this
+    // inline assembly as conservatively clobbering all caller-saved registers,
+    // forcing yield to save any live registers.
+    //
+    // Upon direct observation of this function, the LR is the only register
+    // that is live across the SVC invocation, however, if the yield call is
+    // inlined, it is possible that the LR won't be live at all (commonly seen
+    // for the `loop { yieldk(); }` idiom) or that other registers are live,
+    // thus it is important to let the compiler do the work here.
+    //
+    // According to the AAPCS: A subroutine must preserve the contents of the
+    // registers r4-r8, r10, r11 and SP (and r9 in PCS variants that designate
+    // r9 as v6) As our compilation flags mark r9 as the PIC base register, it
+    // does not need to be saved. Thus we must clobber r0-3, r12, and LR
+    unsafe {
+        asm!(
+            "svc 0"
+            :
+            :
+            : "memory", "r0", "r1", "r2", "r3", "r12", "lr"
+            : "volatile");
+    }
+}
+
+pub fn yieldk_for<F: Fn() -> bool>(cond: F) {
+    while !cond() {
+        yieldk();
+    }
+}
+
+pub fn subscribe<CB: SubscribableCallback>(
+    driver_number: usize,
+    subscribe_number: usize,
+    callback: &mut CB,
+) -> Result<CallbackSubscription, Error> {
+    let return_code = unsafe {
+        subscribe_ptr(
+            driver_number,
+            subscribe_number,
+            c_callback::<CB> as *const _,
+            callback as *mut CB as usize,
+        )
+    };
+
+    return_code.map(|_| CallbackSubscription::new(driver_number, subscribe_number))
+}
+
+extern "C" fn c_callback<CB: SubscribableCallback>(
+    arg0: usize,
+    arg1: usize,
+    arg2: usize,
+    userdata: usize,
+) {
+    let callback = unsafe { &mut *(userdata as *mut CB) };
+    callback.call_rust(arg0, arg1, arg2);
+}
+
+pub unsafe fn subscribe_ptr(
+    major: usize,
+    minor: usize,
+    cb: *const unsafe extern "C" fn(usize, usize, usize, usize),
+    ud: usize,
+) -> Result<u32, Error> {
+    let res;
+    asm!("svc 1" : "={r0}"(res)
+                 : "{r0}"(major) "{r1}"(minor) "{r2}"(cb) "{r3}"(ud)
+                 : "memory"
+                 : "volatile");
+    ReturnCode(res).into()
+}
+
+pub unsafe fn command(major: usize, minor: usize, arg1: usize, arg2: usize) -> Result<u32, Error> {
+    let res;
+    asm!("svc 2" : "={r0}"(res)
+                 : "{r0}"(major) "{r1}"(minor) "{r2}"(arg1) "{r3}"(arg2)
+                 : "memory"
+                 : "volatile");
+    ReturnCode(res).into()
+}
+
+pub fn allow(
+    driver_number: usize,
+    allow_number: usize,
+    buffer_to_share: &mut [u8],
+) -> Result<SharedMemory, Error> {
+    let len = buffer_to_share.len();
+    let return_code = unsafe {
+        allow_ptr(
+            driver_number,
+            allow_number,
+            buffer_to_share.as_mut_ptr(),
+            len,
+        )
+    };
```

Comment:
What about using the `?` operator here instead of `map`? Same for the subscription case.